pandas/core/reshape/pivot.py

```diff
@@ -424,42 +425,48 @@ def _convert_by(by):
 
 @Substitution("\ndata : DataFrame")
 @Appender(_shared_docs["pivot"], indents=1)
-def pivot(data: "DataFrame", index=None, columns=None, values=None) -> "DataFrame":
+def pivot(
+    data: "DataFrame",
+    index: Optional[Union[Label, Sequence[Optional[Label]]]] = None,
+    columns: Optional[Union[Label, Sequence[Optional[Label]]]] = None,
+    values: Optional[Union[Label, Sequence[Optional[Label]]]] = None,
+) -> "DataFrame":
     if columns is None:
         raise TypeError("pivot() missing 1 required argument: 'columns'")
```

Comment:
if this is a required kwarg, should the Optional be removed from the signature?  cc @simonjayhawkins not sure how to handle this given that the default value is None

Reply:
There are a few considerations here. 

`Optional` is normally included to represent a keyword argument that defaults to None. Since `index` is a keyword argument, `columns` must be a keyword argument (non-default argument follows default argument). but because Label already includes None (due to an issue with Hashable) the Optional isn't actually required in this case (This is the reason to exclude Optional from aliases in general).

If Optional was omitted, this could add to the confusion.

Since columns should allow None, if a column is named None. 

so the default and check here should probably be _lib.no_default.

 _lib.no_default currently resolves to Any (defined in cython with no stubs) so is compatible with anything.

If we typed _lib.no_default which is currently object() we wouldn't be able to type parameters precisely since object() wouldn't be compatible.  In the future we may wish to define using Enums but this would add to the noise of the function signature from an end user perspective.

from https://www.python.org/dev/peps/pep-0484/#support-for-singleton-types-in-unions

> To allow precise typing in such situations, the user should use the Union type in conjunction with the enum.Enum class provided by the standard library, so that type errors can be caught statically: